cmd/validator/core/validator.go

```diff
@@ -119,7 +129,101 @@ func Start(work <-chan string, processorID ProcessorID, threads int, config Para
+		} else if split[0] == "box" && len(split) == 3 {
+			callProcessorBox(processor, split[1], split[2], config, results)
+		} else {
+			err := fmt.Errorf("invalid resource")
+			results <- resultResourceError(err, input)
+			return
+		}
+
+	}
+}
+
+func callProcessorBox(processor Processor, appid string, boxname string, config Params, results chan<- Result) {
+	boxname = strings.ReplaceAll(boxname, "+", "%2B")
+	boxname = strings.ReplaceAll(boxname, "/", "%2F")
+
+	algodDataURL := fmt.Sprintf("%s/v2/applications/%s/box?name=b64:%s", config.AlgodURL, appid, boxname)
+	indexerDataURL := fmt.Sprintf("%s/v2/applications/%s/box?name=b64:%s", config.IndexerURL, appid, boxname)
+
+	// Fetch algod box outside the retry loop. When the data desynchronizes we'll keep fetching indexer data until it
+	// catches up with the first algod box query.
+	algodData, err := getData(algodDataURL, config.AlgodToken)
+	if err != nil {
+		err = fmt.Errorf("error getting algod data (%s): %w", algodData, err)
+		switch {
+		case strings.Contains(string(algodData), api.ErrBoxNotFound):
+			results <- resultBoxSkip(err, appid, boxname, SkipBoxNotFound)
+		default:
+			results <- resultBoxError(err, appid, boxname)
+		}
+		return
+	}
+
+	// Retry loop.
+	for i := 0; true; i++ {
+		indexerData, err := getData(indexerDataURL, config.IndexerToken)
+		if err != nil {
+			err = fmt.Errorf("error getting indexer data (%s): %w", indexerData, err)
+			switch {
+			case strings.Contains(string(indexerData), api.ErrFailedLookingUpBoxes):
+				results <- resultBoxSkip(err, appid, boxname, SkipBoxFailedLookup)
+			case strings.Contains(string(indexerData), api.ErrWrongAppidFound):
+				results <- resultBoxSkip(err, appid, boxname, SkipBoxWrongAppid)
+			case strings.Contains(string(indexerData), api.ErrMultipleBoxes):
+				results <- resultBoxSkip(err, appid, boxname, SkipBoxMultiple)
+			case strings.Contains(string(indexerData), api.ErrNoBoxesFound):
+				results <- resultBoxSkip(err, appid, boxname, SkipBoxNoBoxes)
+			case strings.Contains(string(indexerData), api.ErrWrongBoxFound):
+				results <- resultBoxSkip(err, appid, boxname, SkipBoxWrongBox)
+			default:
+				results <- resultBoxError(err, appid, boxname)
+			}
+			return
+		}
+
+		result, err := processor.ProcessBox(algodData, indexerData)
+		if err != nil {
+			// If there is an error return immediately and cram the error.
+			results <- Result{
+				Equal:   false,
+				Error:   fmt.Errorf("error processing box %s, %s: %v", appid, boxname, err),
+				Retries: i,
+				Details: &ErrorDetails{
+					Appid:    appid,
+					Boxname:  boxname,
+					Resource: "box",
+				},
+			}
+			return
+		}
+
+		if result.Equal || (i >= config.Retries) {
+			// Return when results are equal, or when finished retrying.
+			result.Retries = i
+			if result.Details != nil {
+				result.Details.Resource = "box"
+				result.Details.Appid = appid
+				result.Details.Boxname = boxname
+			}
+			results <- result
+			return
+		}
```

Comment:
It's a shame we can't get the rounds too. The vast majority of failures I've seen have mismatched rounds, so it's a quick way to filter out false negatives. We'll have to monitor whether or not this ends up being a problem for Boxes, or if it's a 1 in a million thing that we run into because there are millions of accounts.

Thinking out loud: if it does become a problem, and we don't want to add round to the boxes results, this code could run the account processor to get some extra information.

Reply:
yes we could either add the round to the box endpoint or we could sandwich each call to box with a request for the round number, and repeat if any of the four rounds are different.

Reply:
scytale of the opinion that we should add the round number to the box endpoint and also further add a requested round parameter to the endpoint.